# Notebook #4: Cálculo de Rentabilidad

- Con los datos extraídos y el modelo de *machine learning* que se ha entrenado para predecir el precio del alquiler, somos capaces de realizar el cálculo de rentabilidad con ayuda de distintas funciones.

- Importamos las librerías y soportes.

In [432]:
%load_ext autoreload
%autoreload 2

# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import geopandas as gpd
import numpy as np
import pickle
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los dfFrames

from tqdm import tqdm

# Ignorar los warnings
# -----------------------------------------------------------------------
import warnings
warnings.filterwarnings('ignore')

# Librería para el acceso a variables y funciones
# -----------------------------------------------------------------------
import sys
sys.path.append("../")
from src import soporte_rentabilidad as sr
from src import soporte_yolo as sy
from src import soporte_scoring as ss
from src import soporte_blip as sb
from src import soporte_mongo as sm

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


- Como primer paso, definimos una variable con las rutas de los transformers: el encoder, el scaler y el modelo Random Forest.

- También vamos a conectarnos a la base de datos de Mongo para obtener el los datos correspondientes a viviendas en venta.

- Utilizaremos la función `predecir_alquiler` para aplicar el modelo a la totalidad de viviendas en venta, que devuelve el dataframe con una nueva columna 'alquiler_predicho'. Sus únicos argumentos son el archivo pickle con las viviendas en venta y la variable con las rutas de los transformers.

In [522]:
paths_transformers = ["../transformers/target_encoder.pkl", "../transformers/scaler.pkl", "../transformers/model.pkl"]
bd = sm.conectar_a_mongo("ProyectoRentabilidad")
gdf = sm.importar_a_geodataframe(bd, "venta")
gdf = sr.predecir_alquiler(gdf, paths_transformers)

- Comprobamos el dataframe.

- Para cada vivienda, tenemos una columna que corresponde a una lista con las urls de las imágenes. Para poder identificar cuál corresponde a cada estancia, utilizaremos el modelo YOLOv11, en su versión x y de clasificación, por cuanto no nos interesa determinar la ubicación ni la forma de los objetos en las imágenes, solamente su existencia.

- Con la función `identificar_urls_habitaciones`, se añaden etiquetas a la cocina y al baño de cada anuncio. Para optimizar el tiempo de procesamiento, cuando las ha encontrado, no continua con el resto de imágenes, sino que salta al siguiente anuncio.

- El resultado es un dataframe con nuevas columnas para la URL del baño y de la cocina. Cuando uno de los dos valores es nulo, elimina esa fila. El return también incluye un dataframe con todas las detecciones de las imágenes procesadas, como una forma de poder validar los resultados.

116 misses

In [523]:
gdf, detecciones = sy.identificar_urls_habitaciones(gdf, 'urls_imagenes', drop_nulls=True)

 33%|███▎      | 116/352 [36:30<1:06:39, 16.95s/it] 

Error processing https://img4.idealista.com/blur/WEB_LISTING-M/0/id.pro.es.image.master/f3/c0/22/1263973405.webp: HTTPSConnectionPool(host='img4.idealista.com', port=443): Read timed out. (read timeout=10)
Error processing https://img4.idealista.com/blur/WEB_LISTING-M/0/id.pro.es.image.master/a9/7b/cd/1263973421.webp: HTTPSConnectionPool(host='img4.idealista.com', port=443): Read timed out. (read timeout=10)


100%|██████████| 352/352 [1:13:56<00:00, 12.61s/it]

Se han eliminado 117 filas donde 'url_cocina' o 'url_banio' son None o NaN.


- Vamos a comprobar el resultado del filtrado.

- Además del dataframe con las imágenes filtradas, tenemos un segundo dataframe como retorno, que almacena los elementos que se han detectado en cada imagen, la URL y si ha sido etiquetado como cocina o baño. Esto nos permite realizar generar una métrica sobre la efectividad de YOLO.

- De las 365 viviendas en venta, se han encontrado imágenes de baño y cocina para 266. Dado que el cliente objetivo del proyecto no es quién se dedica a reformas, sin poder conocer el estado de estas estancias, el cálculo de rentabilidad puede cambiar significativamente por ser los espacios que cuesta más dinero reformar.

- Por ese motivo, las eliminaremos del conjunto de datos con el parámetro *drop nulls*.

In [525]:
gdf.shape[0]

265

In [526]:
detecciones[detecciones['habitación']=='kitchen'].head(5)

url  \
2   https://img4.idealista.com/blur/WEB_LISTING-M/...   
7   https://img4.idealista.com/blur/WEB_LISTING-M/...   
13  https://img4.idealista.com/blur/WEB_LISTING-M/...   
14  https://img4.idealista.com/blur/WEB_LISTING-M/...   
16  https://img4.idealista.com/blur/WEB_LISTING-M/...   

                                          detecciones habitación  
2   [microwave, refrigerator, wardrobe, dishwasher...    kitchen  
7   [microwave, washbasin, refrigerator, washer, s...    kitchen  
13  [microwave, washer, stove, dishwasher, washbasin]    kitchen  
14  [washer, microwave, washbasin, refrigerator, s...    kitchen  
16  [washer, microwave, space_heater, dishwasher, ...    kitchen

- Al comprobar las URLs de forma manual utilizando una muestra aleatoria, se confirma una accuracy de entorno al 90% en el etiquetado.

In [527]:
for value in detecciones[detecciones['habitación']=='kitchen']['url'].sample(10):
    print(value)

https://img4.idealista.com/blur/WEB_LISTING-M/0/id.pro.es.image.master/69/26/e9/1278654596.webp
https://img4.idealista.com/blur/WEB_LISTING-M/0/id.pro.es.image.master/d2/f1/46/1300762768.webp
https://img4.idealista.com/blur/WEB_LISTING-M/90/id.pro.es.image.master/e2/c2/66/1284302405.webp
https://img4.idealista.com/blur/WEB_LISTING-M/90/id.pro.es.image.master/27/0a/a8/1159363025.webp
https://img4.idealista.com/blur/WEB_LISTING-M/0/id.pro.es.image.master/f9/90/42/1245503450.webp
https://img4.idealista.com/blur/WEB_LISTING-M/0/id.pro.es.image.master/ff/30/25/1295857292.webp
https://img4.idealista.com/blur/WEB_LISTING-M/0/id.pro.es.image.master/19/5a/4d/1300123246.webp
https://img4.idealista.com/blur/WEB_LISTING-M/0/id.pro.es.image.master/7d/20/9c/1289004786.webp
https://img4.idealista.com/blur/WEB_LISTING-M/0/id.pro.es.image.master/59/1a/b9/1300629581.webp
https://img4.idealista.com/blur/WEB_LISTING-M/0/id.pro.es.image.master/ca/f7/72/918499944.webp


- Como una capa de verificación adicional, vamos a utilizar el modelo BLIP de Salesforce, que crea descripciones de las imágenes, en este caso, de las URLs identificadas por YOLO. Contamos luego el número de ocurrencias de 'kitchen' o 'bathroom'.

In [528]:
captions_cocinas = sb.generar_descripciones(gdf, 'url_cocina')
sb.contar_palabras(captions_cocinas, 'kitchen')

265it [03:50,  1.15it/s]


233

In [510]:
captions_banios = sb.generar_descripciones(gdf, 'url_banio')
sb.contar_palabras(captions_banios, 'bathroom')

266it [04:24,  1.01it/s]


250

- De las 266 imágenes identificadas como cocinas por YOLO, BLIP ha coincidido en 234, un 88%. El dato en el caso de los baños mejora, habiendo encontrado 250 coincidencias, un 94%.

- Conocedores de que, habrá falsos negativos y falsos positivos en ambos modelos, el alto porcentaje de coincidencias en ambos casos nos indica que la fiabilidad del primer cribado es alta.

- Una observación adicional a lo detectado en el EDA es que, aquellos anuncios en que hemos rellenado el valor nulo del anunciante con 'ND', corresponden generalmente a particulares, lo cual es relevante porque podría significar entorno a un 3% de ahorro en el valor de compra.

In [529]:
gdf[gdf['anunciante']=='ND'].sample(2)

codigo    precio  precio_por_zona  tipo exterior planta ascensor  \
234  102216234  144900.0           1932.0  piso     True      3    False   
66   106918117  134000.0           2436.0  piso     True      4    False   

     tamanio  habitaciones  banios aire_acondicionado trastero terraza  patio  \
234     75.0             4       1              False    False    True  False   
66      55.0             2       1               True    False   False  False   

    estado                direccion  \
234   good       calle Portugal, 16   
66    good  calle del Río Duero, 30   

                                           descripcion anunciante   contacto  \
234  Piso ubicado en la zona universitaria, en CALL...         ND  652561233   
66   Esta encantadora y acogedora vivienda, con un ...         ND         ND   

                                         urls_imagenes     distrito  \
234  ['https://img4.idealista.com/blur/WEB_LISTING-...     Delicias   
66   ['https://img4.idealista.com/blur/WEB_LISTING-...  La Almozara   

                      geometry  alquiler_predicho  \
234  POINT (-0.90189 41.65036)              913.0   
66   POINT (-0.89903 41.65904)              686.0   

                                            url_cocina  \
234  https://img4.idealista.com/blur/WEB_LISTING-M/...   
66   https://img4.idealista.com/blur/WEB_LISTING-M/...   

                                             url_banio  
234  https://img4.idealista.com/blur/WEB_LISTING-M/...  
66   https://img4.idealista.com/blur/WEB_LISTING-M/...

- El siguiente paso será utilizar la API de Anthropic, y específicamente el modelo Sonnet, para asignar una puntuación del 1 al 10 al estado de la estancia, y una estimación de los metros cuadrados, que es una consideración que ha surgido durante el desarrollo y la fase de pruebas del proyecto.

- El prompt que se utiliza incluye:

<div style="background-color:rgb(149, 222, 237); padding: 10px; border-left: 6px solid #000080; color: black; border-radius: 10px;">
You are an AI image analysis system specialized in evaluating property conditions. Your task is to analyze multiple property images in batch and provide a precise evaluation.
Instructions for Analysis:

Analyze each pair of images:

Image 1: Kitchen
Image 2: Bathroom

Evaluation Criteria:

Ratings: Whole numbers 1-5
1: Very poor (complete renovation required)
2: Poor (major renovations needed)
3: Fair (some renovations needed)
4: Good (minor improvements required)
5: Excellent (no renovations needed)

Sizes: Whole numbers in square meters (m²)
</div>

- Como puede verse en el prompt, el resultado de la consulta son 4 valores separados por comas, con la función `analizar_propiedades`, separaremos esos resultados y los asignaremos a 4 nuevas columnas del dataframe. Para tener consultas más eficientes, se enviarán en lotes, o *batches*, que corresponden a grupos de viviendas. Este batch más el dataframe con las URLs identificadas son los parámetros de la función.

In [530]:
df_sample = gdf.head(3)

In [532]:
df_sample, resultados = ss.analizar_propiedades(df_sample, batch=3)

Procesando lotes:   0%|          | 0/1 [00:00<?, ?it/s]

In [533]:
df_sample

codigo    precio  precio_por_zona  tipo exterior planta ascensor  \
0  104792745  149900.0           1180.0  piso     True      4     True   
1  106086132  139900.0           1686.0  piso     True      3     True   
2  105844791  149900.0           1162.0  piso     True      4     True   

   tamanio  habitaciones  banios aire_acondicionado trastero terraza  patio  \
0    127.0             3       2              False    False   False   True   
1     83.0             2       1              False    False    True  False   
2    129.0             3       2              False    False   False  False   

  estado                direccion  \
0   good  carretera de Huesca, 21   
1   good        calle de Portugal   
2   good      carretera de Huesca   

                                         descripcion        anunciante  \
0  Junto a la Academia General Militar, en carret...       FINCAS RUIZ   
1  Financiación hipoteca 100% ¡oportunidad única ...  ciz inmobiliaria   
2  Financiación hipoteca 100%. ¡oportunidad única...  ciz inmobiliaria   

    contacto                                      urls_imagenes  \
0  876210884  ['https://img4.idealista.com/blur/WEB_LISTING-...   
1  876210272  ['https://img4.idealista.com/blur/WEB_LISTING-...   
2  876210272  ['https://img4.idealista.com/blur/WEB_LISTING-...   

         distrito                   geometry  alquiler_predicho  \
0  Distrito Rural   POINT (-0.86935 41.6973)             1180.0   
1        Delicias  POINT (-0.90232 41.65202)              761.0   
2  Distrito Rural     POINT (-0.87 41.69848)             1186.0   

                                          url_cocina  \
0  https://img4.idealista.com/blur/WEB_LISTING-M/...   
1  https://img4.idealista.com/blur/WEB_LISTING-M/...   
2  https://img4.idealista.com/blur/WEB_LISTING-M/...   

                                           url_banio puntuacion_cocina  \
0  https://img4.idealista.com/blur/WEB_LISTING-M/...                 4   
1  https://img4.idealista.com/blur/WEB_LISTING-M/...                 4   
2  https://img4.idealista.com/blur/WEB_LISTING-M/...                 4   

  puntuacion_banio mts_cocina mts_banio  
0                5         20         5  
1                4         18         6  
2                4         16         4

In [334]:
gdf.to_pickle("../data/transformed/final_tags.pkl")

In [335]:
df_importado = pd.read_pickle("../data/transformed/final_tags.pkl")
df_importado.shape[0]

278

In [534]:
sr.calcular_rentabilidad_inmobiliaria_wrapper(gdf, porcentaje_entrada=0.2, coste_reformas=10000, comision_agencia=5000, anios=30, tin=0.03, seguro_vida=0, tipo_irpf=0.37, porcentaje_amortizacion=0.4)

codigo    precio  precio_por_zona  tipo exterior planta ascensor  \
190  106848232   92000.0            989.0  piso     True      2    False   
220  106632479   65000.0           1300.0  piso     True     bj    False   
258  106077006  118000.0           1054.0  piso     True      2    False   
104  107024345   85000.0           1700.0  piso     True      1     True   
65   105470662   69000.0           1150.0  piso     True      3    False   
..         ...       ...              ...   ...      ...    ...      ...   
143   99403051  139500.0           2051.0  piso     True     bj    False   
125  106725265  146500.0           2187.0  piso     True      2     True   
150  106821032  125000.0           2016.0  piso     True      4    False   
117  107005535  150000.0           2727.0  piso     True      1     True   
90   104805364  150000.0           2500.0  piso     True      2     True   

     tamanio  habitaciones  banios aire_acondicionado trastero terraza  patio  \
190     93.0             2       2               True    False   False  False   
220     50.0             2       1              False    False   False  False   
258    112.0             4       2              False    False   False  False   
104     50.0             3       1              False    False    True  False   
65      60.0             2       1              False    False   False  False   
..       ...           ...     ...                ...      ...     ...    ...   
143     68.0             2       1              False    False    True  False   
125     67.0             3       1              False    False   False  False   
150     62.0             2       1              False    False   False  False   
117     55.0             2       1              False    False   False  False   
90      60.0             2       1              False    False    True  False   

             estado                           direccion  \
190            good           calle de Basilio Boggiero   
220            good               calle de Pedro Cubero   
258            good       calle de Basilio Boggiero, 72   
104            good      calle de la Madrina Salinas, 2   
65             good             calle de Mariano Cerezo   
..              ...                                 ...   
143            good        calle de la Madre Sacramento   
125            good                  avenida de Navarra   
150            good                 Grupo Alférez Rojas   
117            good  avenida del Compromiso de Caspe, 1   
90   newdevelopment                       Lastanosa, 17   

                                           descripcion  \
190  Findliving te presenta esta magnifica vivienda...   
220  Propiedades Zaragoza lanza esta increible vivi...   
258  Precioso y luminoso piso de 4 habitaciones, to...   
104  En pleno barrio de Las Fuentes se vende una es...   
65   ATENCIÓN INVERSORES! CASCO HISTÓRICO, EN PERFE...   
..                                                 ...   
143  Vivienda con terraza a la venta en la Calle Ma...   
125  ¿Buscas un piso amplio, reformado y con todas ...   
150  Lassira inmobiliaria vende una vivienda en una...   
117  Desde Escala 21 le ofrecemos este fantástico p...   
90   Descubre una oportunidad única en la promoción...   

                  anunciante   contacto  \
190  Findliving Real Estate   876210931   
220     Propiedades Zaragoza  876210746   
258                       ND         ND   
104    De Fondo Inmobiliario  876210713   
65         Fincas San Carlos  876210479   
..                       ...        ...   
143               Espazio ZG  876210097   
125          Doctor Property  876210316   
150     Lassira inmobiliaria  876210298   
117        Escala21 Asesores  876210445   
90                DOMICILIUM  876210234   

                                         urls_imagenes         distrito  \
190  ['https://img4.idealista.com/blur/WEB_LISTING-...  Casco Histórico   
220  ['https://img4.idealista.com/blur/WEB_LISTIN